### DISCUSSION TITLE
How to encrypt API keys input from users in the database

### DISCUSSION THREAD

**Discussion by aquibbaig** [supabase/supabase#22583] [Questions]
I'm using OpenAI provider in my app and I use the "bring your own keys" policy where users can enter their API keys to use the service. How do I store these keys in an encrypted or hashed format, but it is decrypted properly and sent to the client? 

We anyways have RLS so no other user can get someone else's integration so I think we should be fine in terms of security.

**Comment by GaryAustin1** [comment]
https://supabase.com/docs/guides/database/extensions/pgsodium
You can encrypt a column in a table.

**Comment by ani-sha** [comment]
Hi @GaryAustin1, I tried implementing this. However, I am having issues while inserting any record into the table.
```
ERROR:  42501: permission denied for table key
CONTEXT:  SQL statement "SELECT *                     FROM pgsodium.decrypted_key v
  WHERE id = key_uuid AND key_type = 'aead-det'"
PL/pgSQL function pgsodium.crypto_aead_det_encrypt(bytea,bytea,uuid,bytea) line 5 at SQL statement
PL/pgSQL function api_encrypt_secret_api_key() line 3 at assignment
```
I tried following some discussions around permission issues [here](https://github.com/supabase/supabase/issues/12952) to grant access to these functions.
```
GRANT EXECUTE ON FUNCTION pgsodium.crypto_aead_det_encrypt (bytea, bytea, uuid, bytea) TO authenticated;
GRANT EXECUTE ON FUNCTION pgsodium.crypto_aead_det_decrypt (bytea, bytea, uuid, bytea) TO authenticated;
``` 
These didn't work as well. Any other lead on this?
